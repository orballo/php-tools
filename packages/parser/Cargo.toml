[package]
name = "parser"
version.workspace = true
authors.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
rowan.workspace = true
serde.workspace = true
serde_json.workspace = true
nom.workspace = true
lexer.path = "../lexer"
lexer_derive.path = "../lexer_derive"

[dev-dependencies]
insta.workspace = true

[profile.dev.package]
insta.opt-level = 3
similar.opt-level = 3
