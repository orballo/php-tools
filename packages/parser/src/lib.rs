pub mod php_parser;

pub use php_parser::{PHPParser, Rule as PHPRule};
