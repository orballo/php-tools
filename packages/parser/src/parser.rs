// @TODO: the parser core implementation should parse
// the pieces of the scripts, and once the script sections
// are identified, these are the ones that should be parsed
// with the extensions
//
// The text sections could potentially accept extensions to parse HTML, etc.

pub struct Parser {}

impl Parser {
    pub fn builder() -> ParserBuilder {
        ParserBuilder::new()
    }
}

// @TODO: We need to implement a Token struct.
// This struct is going to be used to store token kinds
// with their name, span, and &str to source.
//
// We need to do the lexing first, and if possible
// concurrently over multiple files.
//
// Parsing should be done after lexing, and only using
// the tokens from the lexer's output.
//
// Each feature/token of the language will be
// in a package called extension.
//
// The individual extensions can implement:
// - Token definition and parser (Optional)
// - Expresion definition and parser (Optional)
// - Statement definition and parser (Optional)
// - Bytecode instruction definition (Optional)
// - Bytecode instruction parser (Optional)
//
pub struct ParserBuilder {
    tokens: Vec<Token>,
    parsers: Vec<Box<dyn SubParser>>,
}

impl ParserBuilder {
    pub fn new() -> Self {
        Self {
            tokens: Vec::new(),
            parsers: Vec::new(),
        }
    }

    pub fn parser() -> Self {
        todo!()
    }

    pub fn parsers() -> Self {
        todo!()
    }
}

pub trait SubParser {}
