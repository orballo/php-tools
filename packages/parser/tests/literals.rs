use insta::{assert_yaml_snapshot, glob, with_settings};
use parser::{PHPParser, PHPRule};
use pest::iterators::Pairs;
use pest::Parser;
use serde::Serialize;
use std::fs;

#[derive(Debug, Clone, Serialize)]
struct Comment {
    content: String,
    starting_line: usize,
    ending_line: usize,
}

impl Default for Comment {
    fn default() -> Self {
        Self {
            content: String::new(),
            starting_line: 0,
            ending_line: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
struct Node {
    name: String,
    content: String,
    line: usize,
    length: usize,
}

#[derive(Debug)]
struct Extractor {
    comment: Comment,
    nodes: Vec<Node>,
}

impl Default for Extractor {
    fn default() -> Self {
        Self {
            comment: Comment::default(),
            nodes: Vec::new(),
        }
    }
}

impl Extractor {
    fn extract_nodes(
        &mut self,
        pairs: Pairs<PHPRule>,
        rule: PHPRule,
    ) -> Vec<Node> {
        for pair in pairs {
            match pair.as_rule() {
                // Parse comment to extract test name.
                PHPRule::SINGLE_LINE_COMMENT => {
                    self.comment.content = pair
                        .as_str()
                        .trim_matches(|c| c == '/' || c == ' ')
                        .to_string();
                    self.comment.starting_line =
                        pair.as_span().start_pos().line_col().0;
                    self.comment.ending_line =
                        pair.as_span().end_pos().line_col().0;
                }
                // Extract the node information and add test name
                // if there is a comment one line above.
                _ if pair.as_rule() == rule => {
                    let content = pair.as_str().to_string();
                    let line = pair.as_span().start_pos().line_col().0;
                    let length = pair.as_span().end() - pair.as_span().start();

                    let name = if self.comment.ending_line + 1 == line {
                        self.comment.content.clone()
                    } else {
                        String::new()
                    };

                    self.nodes.push(Node {
                        name,
                        content,
                        line,
                        length,
                    });
                }
                // For any other rule we recursively call this function.
                _ => {
                    self.extract_nodes(pair.into_inner(), rule);
                }
            }
        }

        self.nodes.clone()
    }
}

#[test]
fn single_quoted_string_literal() {
    glob!("fixtures/single_quoted_string_literals.php", |path| {
        let file = fs::read_to_string(path).unwrap();
        let pairs = PHPParser::parse(PHPRule::SCRIPT, &file)
            .expect("Something went wrong parsing the file");
        let literals = Extractor::default()
            .extract_nodes(pairs, PHPRule::SINGLE_QUOTED_STRING_LITERAL);

        with_settings!({prepend_module_to_snapshot => false}, {
            assert_yaml_snapshot!(literals);
        })
    });
}
