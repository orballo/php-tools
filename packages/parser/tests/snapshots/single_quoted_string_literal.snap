---
source: packages/parser/tests/literals.rs
expression: literals
input_file: packages/parser/tests/fixtures/single_quoted_string_literals.php
---
- name: Simple string
  content: "'This is a single-quoted string.'"
  line: 4
  length: 33
- name: Numeric string
  content: "'1234567890'"
  line: 6
  length: 12
- name: Special characters
  content: "'Special characters: \"!@#$%^&*()_+-=[]{}|;:,.<>/?'"
  line: 8
  length: 50
- name: Escape sequences
  content: "'Escape sequences: \\n \\t'"
  line: 10
  length: 25
- name: Quotes inside
  content: "'Quotes inside: \\'single-quoted\\''"
  line: 12
  length: 34
- name: Multiline string
  content: "'Multiline\nstring'"
  line: 14
  length: 18
- name: Unicode string
  content: "'Unicode characters: \\u{1F600} \\u{1F34E}'"
  line: 17
  length: 41
- name: Escaped backslash
  content: "'String with backslash at the end: \\\\'"
  line: 19
  length: 38
- name: Empty string
  content: "''"
  line: 21
  length: 2
- name: Indented multiline string
  content: "'Multiline\n            indented'"
  line: 23
  length: 32
